### PR TITLE
Feature/robust transport

### PR DESCRIPTION
**Pull request by fedyiv**

### CODE REVIEW

File: src/commonMain/kotlin/com/dxc/ssi/agent/api/SsiAgentApi.kt [hunk 1]
```diff
@@ -3,21 +3,28 @@ package com.dxc.ssi.agent.api
...
     fun issueCredentialOverConnection(connection: PeerConnection)
+
     //TODO: understand proper signature for this function
     fun requestProofOverConnection(connection: PeerConnection)
     fun getLedgerConnector(): LedgerConnector
     fun getWalletConnector(): WalletConnector
     fun shutdown(force: Boolean)
-    fun getConnections(): Set<PeerConnection>
-    fun disconnect(connection: PeerConnection, force: Boolean)
-    fun disconnectAll(force: Boolean)
```
Comment (kkamyczek) [inline src/commonMain/kotlin/com/dxc/ssi/agent/api/SsiAgentApi.kt:22]: why did you remove method disconnectAll? It seems reasonable to have it.

File: src/commonMain/kotlin/com/dxc/ssi/agent/didcomm/router/MessageRouterImpl.kt [hunk 2]
```diff
@@ -85,6 +104,7 @@ class MessageRouterImpl(
             typeAttribute.contains("/connections/1") -> Route.DidExchange
             typeAttribute.contains("/present-proof/1") -> Route.CredVerifier
             typeAttribute.contains("/trust_ping/1") -> Route.TrustPing
+            typeAttribute.contains("abandon_connection/1.0") -> Route.AbandonConnection
```
Comment (kkamyczek) [inline src/commonMain/kotlin/com/dxc/ssi/agent/didcomm/router/MessageRouterImpl.kt:107]: can syntax be? "/abandon_connection/1.0"

Comment (fedyiv) [reply src/commonMain/kotlin/com/dxc/ssi/agent/didcomm/router/MessageRouterImpl.kt:107]: Good catch! Will fix it

File: src/commonMain/kotlin/com/dxc/ssi/agent/transport/WebSocketTransportImpl.kt [hunk 3]
```diff
@@ -1,31 +1,60 @@
...
 
-        val appSocket = appSocketThreadSafeProvider.provideAppSocket(connection.endpoint.toString())
 
-        appSocket.send(message.payload)
+        var numberOfRetries = 0
+
+        while (true) {
+            try {
+                val appSocket = appSocketThreadSafeProvider.provideAppSocket(connection.endpoint.toString())
+                //TODO: check with Ilya if we need TrustPing explicitly after reconnect or any message would do?
```
[the remark this answers is not in the document]
Comment (fedyiv) [reply src/commonMain/kotlin/com/dxc/ssi/agent/transport/WebSocketTransportImpl.kt:36]: Done

File: src/commonTest/kotlin/com/dxc/ssi/agent/model/AbandonConnectionAnnounceTest.kt [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+
+        println(jsonString)
+
+        val abandonConnectionAnnounce2 = Json.decodeFromString<AbandonConnectionAnnounce>(jsonString)
+
+        assertEquals(abandonConnectionAnnounce, abandonConnectionAnnounce2)
+
+    }
+
+    @Test
```
[the remark this answers is not in the document]
Comment (fedyiv) [reply src/commonTest/kotlin/com/dxc/ssi/agent/model/AbandonConnectionAnnounceTest.kt:30]: Fixed

File: src/commonTest/kotlin/com/dxc/ssi/agent/model/ConnectionTest.kt [hunk 5]
```diff
@@ -11,11 +11,13 @@ class ConnectionTest {
...
         val connection = PeerConnection(
             id = "id",
-            state = "state",
-            invitation = "invitati",
+            state = PeerConnectionState.INVITATION_RECEIVED,
+            invitation = "invitation",
             isSelfInitiated = true,
             peerRecipientKeys = listOf("keys"),
-            endpoint = Url("ws:\\endpoint:111\\ws")
+            endpoint = Url("ws:\\endpoint:111\\ws"),
```
[the remark this answers is not in the document]
Comment (fedyiv) [reply src/commonTest/kotlin/com/dxc/ssi/agent/model/ConnectionTest.kt:18]: See WebSocketTransportImpl. We already have such enum implemented in ktor: URLProtocol.WSS or  URLProtocol.WS

   @OptIn(InternalAPI::class)
    override suspend fun sendMessage(connection: PeerConnection, message: MessageEnvelop) {
        println("Before sending message to endpoint: ${connection.endpoint}")
        if (!(connection.endpoint.protocol == URLProtocol.WS || connection.endpoint.protocol == URLProtocol.WSS))
            throw IllegalArgumentException("Only websockets are supported by WebSocketTransportImpl!")

File: src/jvmMain/resources/log4j.xml [hunk 6]
```diff
@@ -10,7 +10,7 @@
     </appender>
 
     <root>
-        <level value="TRACE"/>
```
[the remark this answers is not in the document]
Comment (fedyiv) [reply src/jvmMain/resources/log4j.xml:13]: done

**Comment by kkamyczek** [comment]
Approved
